File: java/com/epicdragonworld/gameserver/model/items/Inventory.java
package com.epicdragonworld.gameserver.model.items;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.Map;
import java.util.Map.Entry;
import java.util.concurrent.ConcurrentHashMap;
import java.util.logging.Logger;

import com.epicdragonworld.gameserver.managers.DatabaseManager;

/**
 * @author Pantelis Andrianakis
 */
public class Inventory
{
	private static final Logger LOGGER = Logger.getLogger(Inventory.class.getName());
	
	private static final String RESTORE_INVENTORY = "SELECT * FROM character_items WHERE owner=?";
	private static final String DELETE_INVENTORY = "DELETE FROM character_items WHERE owner=?";
	private static final String STORE_ITEM = "INSERT INTO character_items (owner, slot_id, item_id) values (?, ?, ?)";
	
	private final Map<Integer, Integer> _items = new ConcurrentHashMap<>();
	
	public Inventory(String ownerName)
	{
		// Restore information from database.
		try (Connection con = DatabaseManager.getConnection();
			PreparedStatement ps = con.prepareStatement(RESTORE_INVENTORY))
		{
			ps.setString(1, ownerName);
			try (ResultSet rset = ps.executeQuery())
			{
				while (rset.next())
				{
					_items.put(rset.getInt("slot_id"), rset.getInt("item_id"));
				}
			}
		}
		catch (Exception e)
		{
			LOGGER.warning(e.getMessage());
		}
	}
	
	/**
	 * Only used when player exits the game.
	 * @param ownerName
	 */
	public void store(String ownerName)
	{
		// Delete old records.
		try (Connection con = DatabaseManager.getConnection();
			PreparedStatement ps = con.prepareStatement(DELETE_INVENTORY))
		{
			ps.setString(1, ownerName);
			ps.execute();
		}
		catch (Exception e)
		{
			LOGGER.warning(e.getMessage());
		}
		
		// No need to store if item list is empty.
		if (_items.isEmpty())
		{
			return;
		}
		
		// Store new records.
		try (Connection con = DatabaseManager.getConnection();
			PreparedStatement ps = con.prepareStatement(STORE_ITEM))
		{
			for (Entry<Integer, Integer> slot : _items.entrySet())
			{
				ps.setString(1, ownerName);
				ps.setInt(2, slot.getKey());
				ps.setInt(3, slot.getValue());
				ps.addBatch();
			}
			ps.executeBatch();
		}
		catch (Exception e)
		{
			LOGGER.warning(e.getMessage());
		}
		
		// Clear item list.
		_items.clear();
	}
	
	public int getSlot(int slotId)
	{
		return _items.get(slotId);
	}
	
	public void setSlot(int slotId, int itemId)
	{
		_items.put(slotId, itemId);
	}
	
	public void removeSlot(int slotId)
	{
		_items.remove(slotId);
	}
	
	/**
	 * @return a Map that contains all (slotId, itemId) information.
	 */
	public Map<Integer, Integer> getItems()
	{
		return _items;
	}
}

File: EpicDragonWorldServer/Items/Inventory.cs
using MySql.Data.MySqlClient;
using System;
using System.Collections.Generic;
using System.Text;

/**
 * Author: Pantelis Andrianakis
 * Date: May 5th 2019
 */
public class Inventory
{
    private static readonly string RESTORE_INVENTORY = "SELECT * FROM character_items WHERE owner=@owner";
    private static readonly string DELETE_INVENTORY = "DELETE FROM character_items WHERE owner=@owner";
    private static readonly string STORE_ITEM_START = "INSERT INTO character_items VALUES ";
    private readonly Dictionary<int, int> items = new Dictionary<int, int>();

    public Inventory(string ownerName)
    {
        lock (items)
        {
            // Restore information from database.
            try
            {
                MySqlConnection con = DatabaseManager.GetConnection();
                MySqlCommand cmd = new MySqlCommand(RESTORE_INVENTORY, con);
                cmd.Parameters.AddWithValue("owner", ownerName);
                MySqlDataReader reader = cmd.ExecuteReader();
                while (reader.Read())
                {
                    items.Add(reader.GetInt32("slot_id"), reader.GetInt32("item_id"));
                }
                con.Close();
            }
            catch (Exception e)
            {
                LogManager.Log(e.ToString());
            }
        }
    }

    // Only used when player exits the game.
    public void Store(string ownerName)
    {
        // Delete old records.
        try
        {
            MySqlConnection con = DatabaseManager.GetConnection();
            MySqlCommand cmd = new MySqlCommand(DELETE_INVENTORY, con);
            cmd.Parameters.AddWithValue("owner", ownerName);
            cmd.ExecuteNonQuery();
            con.Close();
        }
        catch (Exception e)
        {
            LogManager.Log(e.ToString());
        }

        // No need to store if item list is empty.
        int itemCount = items.Count;
        if (itemCount == 0)
        {
            return;
        }

        // Prepare query.
        StringBuilder query = new StringBuilder(STORE_ITEM_START);
        foreach (KeyValuePair<int, int> item in items)
        {
            query.Append("('");
            query.Append(ownerName);
            query.Append("',");
            query.Append(item.Key);
            query.Append(",");
            query.Append(item.Value);
            query.Append(")");
            query.Append(itemCount-- == 1 ? ";" : ",");
        }
        // Store new records.
        try
        {
            MySqlConnection con = DatabaseManager.GetConnection();
            MySqlCommand cmd = new MySqlCommand(query.ToString(), con);
            cmd.ExecuteNonQuery();
            con.Close();
        }
        catch (Exception e)
        {
            LogManager.Log(e.ToString());
        }

        // Clear item list?
        // items.Clear();
    }

    public int GetSlot(int slotId)
    {
        if (!items.ContainsKey(slotId))
        {
            return 0;
        }
        return items[slotId];
    }

    public void SetSlot(int slotId, int itemId)
    {
        lock (items)
        {
            items.Add(slotId, itemId);
        }
    }

    public void RemoveSlot(int slotId)
    {
        lock (items)
        {
            items.Remove(slotId);
        }
    }

    public Dictionary<int, int> GetItems()
    {
        return items;
    }
}
